### PR TITLE
samples: sensor: update dht_polling to new API

### PR DESCRIPTION
**Pull request by laurenmurphyx64**
Updates dht_polling sample to new sensor API.
    
Depends on #71093 (merged) and #71160 (still waiting) (commits are included in this PR, so this is DNM until both PRs are merged and I can drop the commits)

### CODE REVIEW

File: samples/sensor/dht_polling/src/main.c [hunk 1]
```diff
@@ -34,27 +52,47 @@ int main(void)
...
-							hum.val1, hum.val2 / 10000);
+			uint32_t temp_fit = 0;
+			uint8_t temp_buf[64];
+
+			decoder->decode(buf,
+					(struct sensor_chan_spec) {SENSOR_CHAN_AMBIENT_TEMP, 0},
+					&temp_fit, 1, temp_buf);
+
+			struct sensor_q31_data *temp_data =
+					(struct sensor_q31_data *)temp_buf;
```
Comment (laurenmurphyx64) [inline samples/sensor/dht_polling/src/main.c:81]: insert space (sensor_q31_data *) temp_buf or delete above... wonky parens and on 86

**Comment by laurenmurphyx64** [comment]
Rebased with updates to https://github.com/zephyrproject-rtos/zephyr/pull/71160

File: samples/sensor/dht_polling/src/main.c [hunk 2]
```diff
@@ -32,29 +50,49 @@ int main(void)
...
-				break;
+				printk("%s: sensor_get_decode() failed: %d\n", dev->name, rc);
+				return rc;
 			}
 
-			printk("%16s: temp is %d.%02d °C humidity is %d.%02d %%RH\n",
-							dev->name, temp.val1, temp.val2 / 10000,
-							hum.val1, hum.val2 / 10000);
+			uint32_t temp_fit = 0;
+			uint8_t temp_buf[64];
```
Comment (MaureenHelm) [inline samples/sensor/dht_polling/src/main.c:74]: This seems too large for a single sample? Same question about `hum_buf[64]` below.

File: samples/sensor/dht_polling/src/main.c [hunk 3]
```diff
@@ -8,17 +8,35 @@
...
 /* Support up to 10 temperature/humidity sensors */
 static const struct device *const sensors[] = {LISTIFY(10, DHT_DEVICE, ())};
 
+#define DHT_IODEV(i, _)                                                      \
+	IF_ENABLED(DT_NODE_EXISTS(DHT_ALIAS(i)),                                 \
+		(SENSOR_DT_READ_IODEV(_CONCAT(dht_iodev, i), DHT_ALIAS(i),           \
+		{SENSOR_CHAN_AMBIENT_TEMP, 0},                                       \
+		{SENSOR_CHAN_HUMIDITY, 0})))
+
+LISTIFY(10, DHT_IODEV, (;));
```
Comment (ubieda) [inline samples/sensor/dht_polling/src/main.c:31]: I'm curious if there's a board/sample that has 10 DHT sensors

**Review by ubieda — DISMISSED** [review]
PR Looks good, assuming rebasing it's done.
I've tried it on my end (on a single instance DHT sensor) and it works as expected:

``` console
*** Booting Zephyr OS build v3.6.0-4878-g1c496f6b97f3 ***
       hts221@5f: temp is 24.87 °C humidity is 44.49 RH
       hts221@5f: temp is 25.12 °C humidity is 43.99 RH
       hts221@5f: temp is 25.12 °C humidity is 43.99 RH
       hts221@5f: temp is 25.24 °C humidity is 43.99 RH
       hts221@5f: temp is 25.24 °C humidity is 43.99 RH
       hts221@5f: temp is 25.24 °C humidity is 43.99 RH
       hts221@5f: temp is 25.37 °C humidity is 43.99 RH
       hts221@5f: temp is 25.37 °C humidity is 43.49 RH
       hts221@5f: temp is 25.37 °C humidity is 43.49 RH
       hts221@5f: temp is 25.49 °C humidity is 43.49 RH
       hts221@5f: temp is 25.49 °C humidity is 43.49 RH
       hts221@5f: temp is 25.49 °C humidity is 43.49 RH
       hts221@5f: temp is 25.49 °C humidity is 43.49 RH
       hts221@5f: temp is 25.62 °C humidity is 43.49 RH
       hts221@5f: temp is 25.62 °C humidity is 43.49 RH
       hts221@5f: temp is 25.74 °C humidity is 43.49 RH
```

File: samples/sensor/dht_polling/src/main.c (hunk 3 above)
Comment (teburd) [reply samples/sensor/dht_polling/src/main.c:31]: I suppose a blame could tell us, I know one person had a ton of ADI 1wire to i2c adapters that subsequently ran to temperature sensors. Quite a lot of them in fact.

**Comment by teburd** [comment]
Please rebase on main when you can, the sample changes look good. The buffer size concerns @MaureenHelm has should be addressed with when both #71235 and #73031 are addressed.

The buffer is large because bme280's compensation parameters are included in the buffer today and that's what @laurenmurphyx64 was testing with. We need to resolve #73031 in order to fix this part.

The buffer is large as well because we don't have a way of deciding, at compile time, how big the buffer should be without the metadata #71235 would presumably provide which @yperess is working on.

File: samples/sensor/dht_polling/src/main.c (hunk 3 above)
Comment (ubieda) [reply samples/sensor/dht_polling/src/main.c:31]: Seems like #63016 :smile:

**Comment by laurenmurphyx64** [comment]
> @laurenmurphyx64 Please rebase on main when you can, the sample changes look good. The buffer size concerns @MaureenHelm has should be addressed with when both #71235 and #73031 are addressed.

Rebased, will standby.

File: samples/sensor/dht_polling/src/main.c (hunk 2 above)
Comment (laurenmurphyx64) [reply samples/sensor/dht_polling/src/main.c]: Changed to use sensor_q31_data struct instead of a buffer, take a look

**Review by ubieda — APPROVED** [review]

**Review by teburd — APPROVED** [review]
The remaining task of updating the read buffer size can be done later IMO once we've sorted out how to obtain the metadata needed, which we are working on in #71235

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**